### ISSUE TITLE
Expose number of bytes processed by StreamDeserializer

### ISSUE THREAD

**Issue by dtolnay** [serde-rs/json#183]
Use case from IRC:

> **\<Yorhel>** Hi everyone. Does serde_json support decoding JSON with trailing data?
> **\<Yorhel>** E.g. I'd like to do something in the form of: let (val, data) = serde_json::from_str("{}garbage");
> **\<Yorhel>** With val being the parsed Value and data pointing to the trailing "garbage"

Bincode has a bytes_read() method on their Deserializer that would support this much better than serde_json. You can kind of fake it like this:

```rust
let mut bytes = 0;
let stream = StreamDeserializer::new(s.bytes().map(|b| { bytes += 1; Ok(b) });
/* deserialize a thing */
let garbage = &s[bytes..];
```

But there must be a better way for us to support this.

**Comment by dtolnay** [comment]
Specifically, I think the number we want is the number of bytes so far deserialized into a successful T.

```rust
let data = b"[0] [1] [";

let mut de = serde_json::Deserializer::from_slice(data).into_iter::<Vec<i32>>();
assert_eq!(0, de.byte_pos());

println!("{:?}", de.next()); // [0]
assert_eq!(3, de.byte_pos());

println!("{:?}", de.next()); // [1]
assert_eq!(7, de.byte_pos());

println!("{:?}", de.next()); // error
assert_eq!(8, de.byte_pos());

// if err.is_eof(), can join the remaining data to new data and continue
let remaining = &data[de.byte_pos()..];
```

**Comment by dtolnay** [comment]
Here is a related use case from @vorner that I believe is addressed by byte_offset().

> **\<vorner>** Hello. I'm trying to create a JSONRPC server, so I want to combine tokio and serde_json. The problem is, I'd either need something like from_slice which would tell me how many bytes it used, or StreamDeserializer that would work on top of Futures' Stream instead of an iterator. Is there a way to get something like that?
> **\<oli_obk_>** vorner: can't you tunr a Stream into an Iterator that simply stops on encountering an error?
> **\<oli_obk_>** vorner: it might even store the error inside the iterator itself, so you can obtain it afterwards
> **\<vorner>** oli_obk_: Stops? Like ends?
> **\<vorner>** I don't worry about errors.
> **\<vorner>** The problem is, the data arrive in chunks over network.
> **\<vorner>** So it returns NotReady
> **\<oli_obk_>** ohh...
> **\<vorner>** I need to stop processing that Stream *for now* and resume later when more data arrive.
> **\<vorner>** Which might be in the middle of an object
> **\<oli_obk_>** hmm... don't we have an impl for \`Read\`
> **\<oli_obk_>** vorner: \`from_reader\` might be more up your alley
> **\<vorner>** Yes. But that'd expect the Read to block, not return EWOULDBLOCK (or what is the exact name).
> **\<oli_obk_>** vorner: oh, just return Ok(0) ^^
> **\<oli_obk_>** actually that's probably wrong
> **\<vorner>** That means EOF, doesn't it?
> **\<oli_obk_>** maybe ErrorKind::WouldBlock
> **\<oli_obk_>** https://doc.rust-lang.org/nightly/std/io/enum.ErrorKind.html
> **\<vorner>** Yes, that'd be the one. But if I return that, the StreamReader would give up forever.
> **\<vorner>** But that WouldBlock means only „No data for now, try again later“
> **\<vorner>** The other option is to feed it with a buffer of whatever data I have. But then I'd need to know how much it consumed.
> **\<oli_obk_>** vorner: yea, that's not implemented yet.. I guess blocking is out of the option?
> **\<oli_obk_>** I mean... you could wrap the blocking op in an async
> **\<vorner>** Well, then I'd need to have a thread for each connection.
> **\<oli_obk_>** oh
> **\<dtolnay>** vorner: this might work for getting number of bytes processed by a StreamDeserializer https://github.com/serde-rs/json/issues/183
> **\<vorner>** Now I'm assuming I get one object per line, so I split the input into lines. But it is possible to have a multi-line objects, at least the jsonrpc specs have such examples.
> **\<vorner>** dtolnay: Yes, that one looks like a reasonable workaround.
> **\<dtolnay>** after each successful read from the stream, you would need to save the bytes read and after the first error (which might be because of a cut-off object) you can keep the rest of the data for next time
> **\<vorner>** At least until I get enough time to understand the internals of the StreamDeserializer and make a Stream-based version. But it seems I find bits to improve on every library I touch. Likely because Rust is so young.

**Comment by dtolnay** [comment]
Fixed in #303.
